src/hooks/useUserProfiles.tsx

```diff
@@ -0,0 +1,151 @@
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import { useAppConfig } from './useAppConfig';
+import { useActiveAccount } from './useActiveAccount';
+import { useHttpClient } from './useHttpClient';
+import { compact, uniq } from 'lodash';
+import { useCache } from './CacheProvider';
+import { useUser } from './useUser';
+import { useAuth } from './useAuth';
+
+export type UserProfile = {
+  id: string;
+  profile: { displayName?: string; picture?: string; givenName?: string };
+};
+
+type UserProfilesContextProps = {
+  getProfiles: () => UserProfile[];
+  isLoading?: boolean;
+};
+
+export const UserProfilesContext = createContext<UserProfilesContextProps>({
+  getProfiles: () => [],
+  isLoading: undefined,
+});
+
+export const UserProfilesContextProvider = ({
+  children,
+}: {
+  children?: React.ReactNode;
+}) => {
+  const { accountHeaders, isLoading, account } = useActiveAccount();
+  const { apiClient } = useHttpClient();
+  const { cache } = useCache();
+  const appConfig = useAppConfig();
+  const profiles = useRef<Map<string, UserProfile>>(new Map());
+  const { isLoggedIn } = useAuth();
+  const placeHolderProfile = (userId: string) => ({
+    id: userId,
+    profile: {
+      displayName: userId,
+    },
+  });
+  const messageTiles = appConfig.data?.homeTab?.messageTiles;
+  const userIds = uniq(messageTiles?.flatMap((tile) => tile.userIds));
+  const [profileState, setProfileState] = useState<Map<string, UserProfile>>(
+    new Map(Object.entries(userIds.map(placeHolderProfile))),
+  );
+
+  useEffect(() => {
+    if (!isLoading && isLoggedIn && accountHeaders) {
+      const getProfileFromCache = async (userId: string) => {
+        const cacheKey = `profile/${userId}`;
+        const cachedItem = await cache?.get(cacheKey);
+        if (cachedItem) {
+          const userProfile = JSON.parse(cachedItem) as UserProfile;
+          profiles.current?.set(cacheKey, userProfile);
+        }
+        return cachedItem;
+      };
+
+      const fetchProfileFromApi = async (userId: string) => {
+        const response = await apiClient.request(
+          'GET /v1/account/users/:userId',
+          { userId },
+          {
+            headers: accountHeaders,
+          },
+        );
+
+        if (response.status === 200) {
+          const cacheKey = `profile/${userId}`;
+          cache?.set(cacheKey, JSON.stringify(response.data));
+          profiles.current?.set(cacheKey, response.data);
+        }
+      };
+
+      const fetchUserProfile = async (userId: string) => {
+        const cachedProfile = await getProfileFromCache(userId);
+        if (!cachedProfile) {
+          await fetchProfileFromApi(userId);
+        }
+      };
+
+      const getUserProfilePromises =
+        !isLoading && accountHeaders ? userIds.map(fetchUserProfile) : [];
+
+      Promise.all(getUserProfilePromises).then(() => {
```

Comment:
How does this perform for a user with lots of messages and an empty cache? I wonder if we should do this in chunks in the order they appear on the screen.

Reply:
True, I really should put some protection here to be sure this doesn't cause issues at scale. I can do this in a follow up.

Reply:
+1 yea let's batch this from the get-go - maybe 10-15 users at a time?

Reply:
Now handled internally to react-query